File: android/src/main/kotlin/com/example/qr_scanner_master/QrCodeGenerator.kt
package com.example.qr_scanner_master

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RectF
import android.graphics.Shader
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.qrcode.QRCodeWriter
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import java.util.EnumMap

class QrCodeGenerator {
    
    fun generateQrCode(data: String, options: GenerationOptions): Bitmap? {
        return try {
            val writer = QRCodeWriter()
            val hints = EnumMap<EncodeHintType, Any>(EncodeHintType::class.java)
            
            // Set error correction level
            hints[EncodeHintType.ERROR_CORRECTION] = when (options.errorCorrectionLevel) {
                "LOW" -> ErrorCorrectionLevel.L
                "MEDIUM" -> ErrorCorrectionLevel.M
                "QUARTILE" -> ErrorCorrectionLevel.Q
                "HIGH" -> ErrorCorrectionLevel.H
                else -> ErrorCorrectionLevel.M
            }
            
            hints[EncodeHintType.MARGIN] = options.margin
            
            val bitMatrix = writer.encode(data, BarcodeFormat.QR_CODE, options.size, options.size, hints)
            val width = bitMatrix.width
            val height = bitMatrix.height
            
            val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            
            // Apply basic colors first
            for (x in 0 until width) {
                for (y in 0 until height) {
                    bitmap.setPixel(
                        x, y, 
                        if (bitMatrix[x, y]) options.foregroundColor else options.backgroundColor
                    )
                }
            }
            
            // Apply advanced styling
            var styledBitmap = bitmap
            
            if (options.gradientColors.isNotEmpty()) {
                styledBitmap = applyGradient(styledBitmap, options)
            }
            
            if (options.roundedCorners) {
                styledBitmap = applyRoundedCorners(styledBitmap, options.cornerRadius)
            }
            
            if (options.logoData != null) {
                styledBitmap = embedLogo(styledBitmap, options.logoData, options.logoSizeRatio)
            }
            
            if (options.addBorder) {
                styledBitmap = addBorder(styledBitmap, options.borderWidth, options.borderColor)
            }
            
            styledBitmap
        } catch (e: Exception) {
            null
        }
    }
    
    private fun applyGradient(bitmap: Bitmap, options: GenerationOptions): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val gradientBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(gradientBitmap)
        
        // Create gradient
        val colors = options.gradientColors.toIntArray()
        val gradient = when {
            options.gradientDirection <= 0.25f -> {
                // Horizontal
                LinearGradient(0f, 0f, width.toFloat(), 0f, colors, null, Shader.TileMode.CLAMP)
            }
            options.gradientDirection <= 0.75f -> {
                // Diagonal
                LinearGradient(0f, 0f, width.toFloat(), height.toFloat(), colors, null, Shader.TileMode.CLAMP)
            }
            else -> {
                // Vertical
                LinearGradient(0f, 0f, 0f, height.toFloat(), colors, null, Shader.TileMode.CLAMP)
            }
        }
        
        val paint = Paint().apply {
            shader = gradient
        }
        
        // Draw original bitmap as mask
        canvas.drawBitmap(bitmap, 0f, 0f, null)
        
        // Apply gradient only to foreground pixels
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
        
        return gradientBitmap
    }
    
    private fun applyRoundedCorners(bitmap: Bitmap, cornerRadius: Float): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val roundedBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(roundedBitmap)
        
        val paint = Paint().apply {
            isAntiAlias = true
        }
        
        val rect = RectF(0f, 0f, width.toFloat(), height.toFloat())
        canvas.drawRoundRect(rect, cornerRadius, cornerRadius, paint)
        
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        canvas.drawBitmap(bitmap, 0f, 0f, paint)
        
        return roundedBitmap
    }
    
    private fun embedLogo(bitmap: Bitmap, logoData: ByteArray, logoSizeRatio: Float): Bitmap {
        return try {
            val logoBitmap = android.graphics.BitmapFactory.decodeByteArray(logoData, 0, logoData.size)
            val width = bitmap.width
            val height = bitmap.height
            val logoSize = (width * logoSizeRatio).toInt()
            
            val scaledLogo = Bitmap.createScaledBitmap(logoBitmap, logoSize, logoSize, true)
            val resultBitmap = bitmap.copy(bitmap.config ?: Bitmap.Config.ARGB_8888, true)
            val canvas = Canvas(resultBitmap)
            
            val logoX = (width - logoSize) / 2f
            val logoY = (height - logoSize) / 2f
            
            // Draw white background for logo
            val backgroundPaint = Paint().apply {
                color = Color.WHITE
                isAntiAlias = true
            }
            val backgroundRadius = logoSize / 2f + 10f
            canvas.drawCircle(
                logoX + logoSize / 2f,
                logoY + logoSize / 2f,
                backgroundRadius,
                backgroundPaint
            )
            
            // Draw logo
            canvas.drawBitmap(scaledLogo, logoX, logoY, null)
            
            resultBitmap
        } catch (e: Exception) {
            bitmap
        }
    }
    
    private fun addBorder(bitmap: Bitmap, borderWidth: Int, borderColor: Int): Bitmap {
        val width = bitmap.width + (borderWidth * 2)
        val height = bitmap.height + (borderWidth * 2)
        val borderedBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(borderedBitmap)
        
        // Fill with border color
        canvas.drawColor(borderColor)
        
        // Draw original bitmap in center
        canvas.drawBitmap(bitmap, borderWidth.toFloat(), borderWidth.toFloat(), null)
        
        return borderedBitmap
    }
}

data class GenerationOptions(
    val size: Int = 512,
    val errorCorrectionLevel: String = "MEDIUM",
    val foregroundColor: Int = Color.BLACK,
    val backgroundColor: Int = Color.WHITE,
    val margin: Int = 20,
    val logoData: ByteArray? = null,
    val logoSizeRatio: Float = 0.2f,
    val roundedCorners: Boolean = false,
    val cornerRadius: Float = 4f,
    val gradientColors: List<Int> = emptyList(),
    val gradientDirection: Float = 0f,
    val addBorder: Boolean = false,
    val borderWidth: Int = 2,
    val borderColor: Int = Color.BLACK
) {
    companion object {
        fun fromMap(map: Map<String, Any>): GenerationOptions {
            return GenerationOptions(
                size = (map["size"] as? Number)?.toInt() ?: 512,
                errorCorrectionLevel = map["errorCorrectionLevel"] as? String ?: "MEDIUM",
                foregroundColor = (map["foregroundColor"] as? Number)?.toInt() ?: Color.BLACK,
                backgroundColor = (map["backgroundColor"] as? Number)?.toInt() ?: Color.WHITE,
                margin = (map["margin"] as? Number)?.toInt() ?: 20,
                logoData = map["logoData"] as? ByteArray,
                logoSizeRatio = (map["logoSizeRatio"] as? Number)?.toFloat() ?: 0.2f,
                roundedCorners = map["roundedCorners"] as? Boolean ?: false,
                cornerRadius = (map["cornerRadius"] as? Number)?.toFloat() ?: 4f,
                gradientColors = (map["gradientColors"] as? List<*>)?.mapNotNull { 
                    (it as? Number)?.toInt() 
                } ?: emptyList(),
                gradientDirection = (map["gradientDirection"] as? Number)?.toFloat() ?: 0f,
                addBorder = map["addBorder"] as? Boolean ?: false,
                borderWidth = (map["borderWidth"] as? Number)?.toInt() ?: 2,
                borderColor = (map["borderColor"] as? Number)?.toInt() ?: Color.BLACK
            )
        }
    }
}

File: ios/Classes/QrCodeGenerator.swift
import UIKit
import CoreImage

class QrCodeGenerator {
    
    func generateQrCode(data: String, options: GenerationOptions) -> Data? {
        guard let filter = CIFilter(name: "CIQRCodeGenerator") else { return nil }
        
        let data = data.data(using: .utf8)
        filter.setValue(data, forKey: "inputMessage")
        
        // Set error correction level
        let correctionLevel: String
        switch options.errorCorrectionLevel {
        case "LOW":
            correctionLevel = "L"
        case "MEDIUM":
            correctionLevel = "M"
        case "QUARTILE":
            correctionLevel = "Q"
        case "HIGH":
            correctionLevel = "H"
        default:
            correctionLevel = "M"
        }
        filter.setValue(correctionLevel, forKey: "inputCorrectionLevel")
        
        guard let outputImage = filter.outputImage else { return nil }
        
        // Scale the image to desired size
        let scaleX = CGFloat(options.size) / outputImage.extent.width
        let scaleY = CGFloat(options.size) / outputImage.extent.height
        let scaledImage = outputImage.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))
        
        // Create bitmap context
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue)
        
        guard let context = CGContext(
            data: nil,
            width: options.size,
            height: options.size,
            bitsPerComponent: 8,
            bytesPerRow: 0,
            space: colorSpace,
            bitmapInfo: bitmapInfo.rawValue
        ) else { return nil }
        
        // Fill background
        context.setFillColor(UIColor(rgb: options.backgroundColor).cgColor)
        context.fill(CGRect(x: 0, y: 0, width: options.size, height: options.size))
        
        // Create CI context and render
        let ciContext = CIContext()
        guard let cgImage = ciContext.createCGImage(scaledImage, from: scaledImage.extent) else { return nil }
        
        // Apply foreground color
        context.setBlendMode(.multiply)
        context.setFillColor(UIColor(rgb: options.foregroundColor).cgColor)
        context.fill(CGRect(x: 0, y: 0, width: options.size, height: options.size))
        
        context.setBlendMode(.destinationIn)
        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: options.size, height: options.size))
        
        guard var finalImage = context.makeImage() else { return nil }
        
        // Apply styling
        var styledImage = UIImage(cgImage: finalImage)
        
        if !options.gradientColors.isEmpty {
            styledImage = applyGradient(to: styledImage, options: options) ?? styledImage
        }
        
        if options.roundedCorners {
            styledImage = applyRoundedCorners(to: styledImage, radius: options.cornerRadius) ?? styledImage
        }
        
        if let logoData = options.logoData {
            styledImage = embedLogo(in: styledImage, logoData: logoData, sizeRatio: options.logoSizeRatio) ?? styledImage
        }
        
        if options.addBorder {
            styledImage = addBorder(to: styledImage, width: options.borderWidth, color: UIColor(rgb: options.borderColor)) ?? styledImage
        }
        
        return styledImage.pngData()
    }
    
    private func applyGradient(to image: UIImage, options: GenerationOptions) -> UIImage? {
        let size = image.size
        
        UIGraphicsBeginImageContextWithOptions(size, false, image.scale)
        defer { UIGraphicsEndImageContext() }
        
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        
        // Create gradient
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let colors = options.gradientColors.map { UIColor(rgb: $0).cgColor }
        guard let gradient = CGGradient(colorsSpace: colorSpace, colors: colors as CFArray, locations: nil) else { return nil }
        
        // Determine gradient direction
        let startPoint: CGPoint
        let endPoint: CGPoint
        
        if options.gradientDirection <= 0.25 {
            // Horizontal
            startPoint = CGPoint(x: 0, y: size.height / 2)
            endPoint = CGPoint(x: size.width, y: size.height / 2)
        } else if options.gradientDirection <= 0.75 {
            // Diagonal
            startPoint = CGPoint(x: 0, y: 0)
            endPoint = CGPoint(x: size.width, y: size.height)
        } else {
            // Vertical
            startPoint = CGPoint(x: size.width / 2, y: 0)
            endPoint = CGPoint(x: size.width / 2, y: size.height)
        }
        
        // Draw original image as mask
        image.draw(at: .zero)
        
        // Apply gradient with mask
        context.setBlendMode(.sourceIn)
        context.drawLinearGradient(gradient, start: startPoint, end: endPoint, options: [])
        
        return UIGraphicsGetImageFromCurrentImageContext()
    }
    
    private func applyRoundedCorners(to image: UIImage, radius: CGFloat) -> UIImage? {
        let size = image.size
        
        UIGraphicsBeginImageContextWithOptions(size, false, image.scale)
        defer { UIGraphicsEndImageContext() }
        
        let rect = CGRect(origin: .zero, size: size)
        let path = UIBezierPath(roundedRect: rect, cornerRadius: radius)
        path.addClip()
        
        image.draw(in: rect)
        
        return UIGraphicsGetImageFromCurrentImageContext()
    }
    
    private func embedLogo(in image: UIImage, logoData: Data, sizeRatio: CGFloat) -> UIImage? {
        guard let logoImage = UIImage(data: logoData) else { return image }
        
        let size = image.size
        let logoSize = CGSize(width: size.width * sizeRatio, height: size.height * sizeRatio)
        
        UIGraphicsBeginImageContextWithOptions(size, false, image.scale)
        defer { UIGraphicsEndImageContext() }
        
        // Draw original QR code
        image.draw(at: .zero)
        
        // Draw white background for logo
        let logoRect = CGRect(
            x: (size.width - logoSize.width) / 2,
            y: (size.height - logoSize.height) / 2,
            width: logoSize.width,
            height: logoSize.height
        )
        
        let backgroundRect = logoRect.insetBy(dx: -10, dy: -10)
        let backgroundPath = UIBezierPath(ovalIn: backgroundRect)
        UIColor.white.setFill()
        backgroundPath.fill()
        
        // Draw logo
        logoImage.draw(in: logoRect)
        
        return UIGraphicsGetImageFromCurrentImageContext()
    }
    
    private func addBorder(to image: UIImage, width: Int, color: UIColor) -> UIImage? {
        let borderWidth = CGFloat(width)
        let newSize = CGSize(
            width: image.size.width + (borderWidth * 2),
            height: image.size.height + (borderWidth * 2)
        )
        
        UIGraphicsBeginImageContextWithOptions(newSize, false, image.scale)
        defer { UIGraphicsEndImageContext() }
        
        // Fill with border color
        color.setFill()
        UIRectFill(CGRect(origin: .zero, size: newSize))
        
        // Draw original image in center
        let imageRect = CGRect(
            x: borderWidth,
            y: borderWidth,
            width: image.size.width,
            height: image.size.height
        )
        image.draw(in: imageRect)
        
        return UIGraphicsGetImageFromCurrentImageContext()
    }
}

struct GenerationOptions {
    let size: Int
    let errorCorrectionLevel: String
    let foregroundColor: Int
    let backgroundColor: Int
    let margin: Int
    let logoData: Data?
    let logoSizeRatio: CGFloat
    let roundedCorners: Bool
    let cornerRadius: CGFloat
    let gradientColors: [Int]
    let gradientDirection: CGFloat
    let addBorder: Bool
    let borderWidth: Int
    let borderColor: Int
    
    static func fromDictionary(_ dict: [String: Any]) -> GenerationOptions {
        return GenerationOptions(
            size: dict["size"] as? Int ?? 512,
            errorCorrectionLevel: dict["errorCorrectionLevel"] as? String ?? "MEDIUM",
            foregroundColor: dict["foregroundColor"] as? Int ?? 0xFF000000,
            backgroundColor: dict["backgroundColor"] as? Int ?? 0xFFFFFFFF,
            margin: dict["margin"] as? Int ?? 20,
            logoData: (dict["logoData"] as? FlutterStandardTypedData)?.data,
            logoSizeRatio: CGFloat(dict["logoSizeRatio"] as? Double ?? 0.2),
            roundedCorners: dict["roundedCorners"] as? Bool ?? false,
            cornerRadius: CGFloat(dict["cornerRadius"] as? Double ?? 4.0),
            gradientColors: dict["gradientColors"] as? [Int] ?? [],
            gradientDirection: CGFloat(dict["gradientDirection"] as? Double ?? 0.0),
            addBorder: dict["addBorder"] as? Bool ?? false,
            borderWidth: dict["borderWidth"] as? Int ?? 2,
            borderColor: dict["borderColor"] as? Int ?? 0xFF000000
        )
    }
}

extension UIColor {
    convenience init(rgb: Int) {
        let red = CGFloat((rgb >> 16) & 0xFF) / 255.0
        let green = CGFloat((rgb >> 8) & 0xFF) / 255.0
        let blue = CGFloat(rgb & 0xFF) / 255.0
        let alpha = CGFloat((rgb >> 24) & 0xFF) / 255.0
        
        self.init(red: red, green: green, blue: blue, alpha: alpha)
    }
}
